content/en/blog/_posts/2024-04-17-recursive-read-only-mounts.md

```diff
@@ -0,0 +1,97 @@
+---
+layout: blog
+title: 'Kubernetes 1.30: Read-only volume mounts can be finally literally read-only'
+date: 2024-04-17
+slug: recursive-read-only-mounts
+---
+
+**Author:** Akihiro Suda (NTT)
+
+Kubernetes 1.30 added an experimental support for "recursive read-only mounts".
+
+Prior to this release, read-only volume mounts were surprisingly not really read-only under specific conditions.
+
+### Read-only volume mounts are not really read-only
+
+Volume mounts are widely misunderstood.
+
+The following manifest is expected to make everything under `/mnt` in the containers read-only:
+```yaml
+apiVersion: v1
+kind: Pod
+spec:
+  volumes:
+    - name: mnt
+      hostPath:
+        path: /mnt
+  containers:
+    - volumeMounts:
+        - name: mnt
+          mountPath: /mnt
+          readOnly: true
+```
+
+However, it actually leaves submounts beneath `/mnt` (e.g., NFS or removable media) writable.
+
+### New mount option: recursiveReadOnly
+
+Kubernetes 1.30 added a new mount option `recursiveReadOnly` so as to make submounts recursively read-only.
+
+The option can be enabled as follows:
+```yaml
+apiVersion: v1
+kind: Pod
+spec:
+  volumes:
+    - name: mnt
+      hostPath:
+        path: /mnt
+  containers:
+    - volumeMounts:
+        - name: mnt
+          mountPath: /mnt
+          readOnly: true
+          # NEW
+          # Possible values are `Enabled`, `IfPossible`, and `Disabled`.
+          # Needs to be specified in conjunction with `readOnly: true`.
+          recursiveReadOnly: Enabled
+```
+
+This is implemented by applying the `MOUNT_ATTR_RDONLY` attribute with the `AT_RECURSIVE` flag
+using [`mount_setattr(2)`](https://man7.org/linux/man-pages/man2/mount_setattr.2.html).
+
+### Availability
+
+To enable `recursiveReadOnly` mounts, the following components have to be used:
+
+* Kubernetes: v1.30 or later, with the `RecursiveReadOnlyMounts`
+  [feature gate](/docs/reference/command-line-tools-reference/feature-gates/) enabled.
+  As of v1.30, the gate is marked as alpha.
+
+* CRI runtime:
+  * containerd: v2.0 or later
```

Comment:
Can mention that CRI-O 1.30 will have this support.

Reply:
- Is this something that is guaranteed to arrive as part of the v1.30 release of Kubernetes, and therefore to have already happened by the date of publication?
- Are we committing on behalf of Kubernetes that CRI-O 1.30 has this support? CRI-O is a separate project.

We [avoid making statements about the future](https://kubernetes.io/docs/contribute/style/style-guide/#avoid-statements-about-the-future) in tech docs, and I'd extend this to any post-release blog articlle where we're writing on behalf of Kubernetes.

Reply:
Let me keep this as-is until CRI-O actually implement it